hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -610,8 +610,18 @@ class Emit[C](
       case Void() =>
         Code._empty
 
-      case Begin(xs) =>
-        xs.foreach(x => emitVoid(x))
+      case x@Begin(xs) =>
+        if (!ctx.inLoopCriticalPath.contains(x) && xs.forall(x => !ctx.inLoopCriticalPath.contains(x))) {
```

Comment:
method splitter doesn't do well with lots of flat children -- splitting manually does better.